### ISSUE TITLE
Bug: Possible precision loss when using KV cache

### ISSUE THREAD

**Issue by uwu-420** [ggerganov/llama.cpp#7859]
### What happened?

Hi all,

When using the KV cache I'm getting different results compared to the initial run, where the KV cache was empty.
See the following minimal example. I'm using [GPT2](https://huggingface.co/openai-community/gpt2) and converted it with `convert-hf-to-gguf.py`

```cpp
#include <iostream>
#include "llama.h"
#include "common.h"

int main(int argc, char ** argv) {
    llama_backend_init();
    llama_model_params model_params = llama_model_default_params();
    llama_model * model = llama_load_model_from_file("models/gpt2/ggml-model-f16.gguf", model_params);
    llama_context_params ctx_params = llama_context_default_params();
    ctx_params.seed = 42;
    llama_context * ctx = llama_new_context_with_model(model, ctx_params);
    const int inp_size = 256;
    llama_batch batch = llama_batch_init(inp_size, 0, 1);
    std::vector<llama_token> token_ids;
    for (int i = 0; i < inp_size; ++i) {
        llama_token token_id = 1000 + i;
        token_ids.push_back(token_id);
        llama_batch_add(batch, token_id, i, { 0 }, i == (inp_size - 1));
    }
    llama_decode(ctx, batch);
    auto n_vocab = llama_n_vocab(model);
    auto * logits = llama_get_logits_ith(ctx, inp_size - 1);
    std::vector<float> logits1;
    for (int i = 0; i < n_vocab; ++i) {
        logits1.push_back(logits[i]);
    }
    llama_batch_clear(batch);
    llama_batch_add(batch, token_ids[inp_size - 1], inp_size - 1, { 0 }, true);
    llama_decode(ctx, batch);
    logits = llama_get_logits_ith(ctx, 0);
    std::vector<float> logits2;
    for (int i = 0; i < n_vocab; ++i) {
        logits2.push_back(logits[i]);
    }
    std::cout << logits1[0] << " " << logits2[0] << std::endl;
}
```

which prints out `-74.4315 -77.7947` on my machine.

### Name and Version

version: 3121 (1f0dabda)
built with Apple clang version 15.0.0 (clang-1500.0.40.1) for arm64-apple-darwin23.5.0

### What operating system are you seeing the problem on?

Mac

### Relevant log output

```shell
llama_model_loader: loaded meta data with 17 key-value pairs and 149 tensors from models/gpt2/ggml-model-f16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = gpt2
llama_model_loader: - kv   1:                               general.name str              = gpt2
llama_model_loader: - kv   2:                           gpt2.block_count u32              = 12
llama_model_loader: - kv   3:                        gpt2.context_length u32              = 1024
llama_model_loader: - kv   4:                      gpt2.embedding_length u32              = 768
llama_model_loader: - kv   5:                   gpt2.feed_forward_length u32              = 3072
llama_model_loader: - kv   6:                  gpt2.attention.head_count u32              = 12
llama_model_loader: - kv   7:          gpt2.attention.layer_norm_epsilon f32              = 0.000010
llama_model_loader: - kv   8:                          general.file_type u32              = 1
llama_model_loader: - kv   9:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  10:                         tokenizer.ggml.pre str              = gpt-2
llama_model_loader: - kv  11:                      tokenizer.ggml.tokens arr[str,50257]   = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  12:                  tokenizer.ggml.token_type arr[i32,50257]   = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  13:                      tokenizer.ggml.merges arr[str,50000]   = ["Ġ t", "Ġ a", "h e", "i n", "r e",...
llama_model_loader: - kv  14:                tokenizer.ggml.bos_token_id u32              = 50256
llama_model_loader: - kv  15:                tokenizer.ggml.eos_token_id u32              = 50256
llama_model_loader: - kv  16:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   99 tensors
llama_model_loader: - type  f16:   50 tensors
llm_load_vocab: special tokens cache size = 1
llm_load_vocab: token to piece cache size = 0.3060 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = gpt2
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 50257
llm_load_print_meta: n_merges         = 50000
llm_load_print_meta: n_ctx_train      = 1024
llm_load_print_meta: n_embd           = 768
llm_load_print_meta: n_head           = 12
llm_load_print_meta: n_head_kv        = 12
llm_load_print_meta: n_layer          = 12
llm_load_print_meta: n_rot            = 64
llm_load_print_meta: n_embd_head_k    = 64
llm_load_print_meta: n_embd_head_v    = 64
llm_load_print_meta: n_gqa            = 1
llm_load_print_meta: n_embd_k_gqa     = 768
llm_load_print_meta: n_embd_v_gqa     = 768
llm_load_print_meta: f_norm_eps       = 1.0e-05
llm_load_print_meta: f_norm_rms_eps   = 0.0e+00
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 3072
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = -1
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 10000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 1024
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 0.1B
llm_load_print_meta: model ftype      = F16
llm_load_print_meta: model params     = 163.04 M
llm_load_print_meta: model size       = 312.70 MiB (16.09 BPW) 
llm_load_print_meta: general.name     = gpt2
llm_load_print_meta: BOS token        = 50256 '<|endoftext|>'
llm_load_print_meta: EOS token        = 50256 '<|endoftext|>'
llm_load_print_meta: LF token         = 128 'Ä'
llm_load_print_meta: EOT token        = 50256 '<|endoftext|>'
llm_load_tensors: ggml ctx size =    0.14 MiB
ggml_backend_metal_log_allocated_size: allocated buffer, size =   312.72 MiB, (  312.80 / 27648.00)
llm_load_tensors: offloading 12 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 13/13 layers to GPU
llm_load_tensors:        CPU buffer size =    76.62 MiB
llm_load_tensors:      Metal buffer size =   312.71 MiB
............................................
llama_new_context_with_model: n_ctx      = 512
llama_new_context_with_model: n_batch    = 512
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 10000.0
llama_new_context_with_model: freq_scale = 1
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M3 Pro
ggml_metal_init: picking default device: Apple M3 Pro
ggml_metal_init: loading '/Users/user/Documents/Repos/llama.cpp/build/bin/default.metallib'
ggml_metal_init: GPU name:   Apple M3 Pro
ggml_metal_init: GPU family: MTLGPUFamilyApple9  (1009)
ggml_metal_init: GPU family: MTLGPUFamilyCommon3 (3003)
ggml_metal_init: GPU family: MTLGPUFamilyMetal3  (5001)
ggml_metal_init: simdgroup reduction support   = true
ggml_metal_init: simdgroup matrix mul. support = true
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: recommendedMaxWorkingSetSize  = 28991.03 MB
llama_kv_cache_init:      Metal KV buffer size =    18.00 MiB
llama_new_context_with_model: KV self size  =   18.00 MiB, K (f16):    9.00 MiB, V (f16):    9.00 MiB
llama_new_context_with_model:        CPU  output buffer size =     0.19 MiB
llama_new_context_with_model:      Metal compute buffer size =   101.16 MiB
llama_new_context_with_model:        CPU compute buffer size =     4.01 MiB
llama_new_context_with_model: graph nodes  = 453
llama_new_context_with_model: graph splits = 2
```

**Comment by ggerganov** [comment]
You have to clear the KV cache after the first batch, otherwise you will end up with a context that is twice as big than intended, with each token present 2 times

**Comment by uwu-420** [comment]
Sorry then, my bad :/

But thanks a lot for explaining :)

**Comment by uwu-420** [comment]
I think I got what you mean and adjusted the example. I compare the results of processing all N tokens of the sequence at once vs. first processing the first N - 1 tokens and then in another step process the last token (where we should make proper use of the cache). The difference between the logits got a lot better, although still somewhat noticeable I'd say. This example gives me `-74.4315 -74.4318`. Anyway, I think that's acceptable. But still FYI @ggerganov

On a related note, isn't it a bit counterintuitive that I have to clear the KV cache? As an API user I'd expect it to not create duplicate entries, especially because I use the same seq_id and seq_pos, which should uniquely identify the cache entry and would allow for avoiding some computation.

```cpp
#include <iostream>
#include "llama.h"
#include "common.h"

int main(int argc, char ** argv) {
    llama_backend_init();
    llama_model_params model_params = llama_model_default_params();
    llama_model * model = llama_load_model_from_file("models/gpt2/ggml-model-f16.gguf", model_params);
    llama_context_params ctx_params = llama_context_default_params();
    ctx_params.seed = 42;
    llama_context * ctx = llama_new_context_with_model(model, ctx_params);

    // process all N tokens at once
    const int inp_size = 256;
    llama_batch batch = llama_batch_init(inp_size, 0, 1);
    std::vector<llama_token> token_ids;
    for (int i = 0; i < inp_size; ++i) {
        llama_token token_id = 1000 + i;
        token_ids.push_back(token_id);
        llama_batch_add(batch, token_id, i, { 0 }, false);
    }
    GGML_ASSERT(batch.n_tokens == inp_size);
    batch.logits[batch.n_tokens - 1] = true;
    llama_decode(ctx, batch);
    auto n_vocab = llama_n_vocab(model);
    auto * logits = llama_get_logits_ith(ctx, inp_size - 1);
    std::vector<float> logits1;
    for (int i = 0; i < n_vocab; ++i) {
        logits1.push_back(logits[i]);
    }

    llama_kv_cache_clear(ctx);

    // process the first N - 1 tokens
    llama_batch_clear(batch);
    for (int i = 0; i < inp_size - 1; ++i) {
        llama_batch_add(batch, token_ids[i], i, { 0 }, false);
    }
    llama_decode(ctx, batch);

    // process the last token
    llama_batch_clear(batch);
    llama_batch_add(batch, token_ids[inp_size - 1], inp_size - 1, { 0 }, true);
    llama_decode(ctx, batch);
    logits = llama_get_logits_ith(ctx, 0);
    std::vector<float> logits2;
    for (int i = 0; i < n_vocab; ++i) {
        logits2.push_back(logits[i]);
    }
    std::cout << logits1[0] << " " << logits2[0] << std::endl;
}
```

**Comment by ggerganov** [comment]
> On a related note, isn't it a bit counterintuitive that I have to clear the KV cache? As an API user I'd expect it to not create duplicate entries, especially because I use the same seq_id and seq_pos, which should uniquely identify the cache entry and would allow for avoiding some computation.

You can avoid the computation of the first N-1 tokens by instead of using `llama_kv_cache_clear()`, use `llama_kv_cache_seq_rm()` to remove the `N`th token and then resubmit it alone. I.e. you can replace the following code with a single `llama_kv_cache_seq_rm()` call:

```c++
    llama_kv_cache_clear(ctx);

    // process the first N - 1 tokens
    llama_batch_clear(batch);
    for (int i = 0; i < inp_size - 1; ++i) {
        llama_batch_add(batch, token_ids[i], i, { 0 }, false);
    }
    llama_decode(ctx, batch);
```
